system/jlib/jmutex.cpp

```diff
@@ -321,7 +321,191 @@ void Monitor::notifyAll()
+        {
+            if (!writeSem.wait(timeout))
+            {
+                cs.enter(); 
+                if (!writeSem.wait(0))
+                {
+                    writeWaiting--;
+                    cs.leave();
+                    return false;
+                }
+                cs.leave();
+            }
+        }
+        else
+            writeSem.wait();
+        //NB: waiting and locks adjusted before the signal occurs.
+    }
+#ifdef _DEBUG
+    exclWriteOwner = GetCurrentThreadId();
+#endif
+    return true;
+}
+
+bool ReadWriteLock::changeToWrite(bool timed, unsigned timeout)
+{
+    cs.enter();
+
+    // invalid use cases, changeToWrite must only be called when read locked.
+    if (writeLocks)
+    {
+        cs.leave();
+        throw makeStringException(9999, "Internal Error: ReadWriteLock::changeToWrite - already write locked");
+    }
+    else if (0 == readLocks)
+    {
+        cs.leave();
+        throw makeStringException(9999, "Internal Error: ReadWriteLock::changeToWrite - not read locked");
+    }
+
+    if (readLocks == 1)
+    {
+        readLocks--;
+        writeLocks++;
+        cs.leave();
+    }
+    else // readLocks>1
+    {
+        readToWriteWaiting++;
+        cs.leave();
+        if (timed)
+        {
+            if (!readToWriteSem.wait(timeout))
+            {
+                cs.enter(); 
+                if (!readToWriteSem.wait(0))
+                {
+                    readToWriteWaiting--;
+                    cs.leave();
+                    return false;
+                }
+                cs.leave();
+            }
+        }
+        else
+            readToWriteSem.wait();
+        //NB: waiting and locks adjusted before the signal occurs.
+    }
+#ifdef _DEBUG
+    exclWriteOwner = GetCurrentThreadId();
+#endif
+    return true;
+}
+
+bool ReadWriteLock::changeToRead()
+{
+    cs.enter();
+    if (!writeLocks) // NB: If the lock is already a read lock then this function will return false.
+    {
+        cs.leave();
+        throw makeStringException(9999, "Internal Error: ReadWriteLock::changeToRead - not write locked");
+
+    }
+    --writeLocks;
+    dbgassertex(!readLocks);
+    readLocks++;
+    cs.leave();
+#ifdef _DEBUG
+    exclWriteOwner = 0;
+#endif
+    return true;
+}
+
+void ReadWriteLock::unlock()
+{
+    cs.enter(); 
+    if (readLocks) // implies this unlock() is paired with a previous read lock
+    {
+        readLocks--;
+        if (readToWriteWaiting && (1 == readLocks))
+        {
+            readToWriteWaiting--;
+            readLocks--;
+            writeLocks++;
+            readToWriteSem.signal();
+        }
+        else if (writeWaiting && (0 == readLocks))
+        {
+            writeWaiting--;
+            writeLocks++;
+            writeSem.signal();
+        }
+    }
+    else // implies this unlock() is paired with a previous write lock
+    {
+        writeLocks--;
+        dbgassertex(writeLocks == 0);
+        if (readWaiting)
```

Comment:
So we are ok with readers always acquiring before any writers ?
Perhaps the check for writers should come first ?  Or we could rotate the order of this check on each unlock ?

Reply:
not sure, but that is not new semantics, ReadWriteLock has always given priority to pending blocked readers over blocked writers.

Worth considering, but as a separate task/discussion/JIRA.